### PR TITLE
Refactoring of SetSubnodeOwner

### PR DESCRIPTION
**Pull request by nxt3d**
This is a complete rewrite of the setSubnodeOwner function. The gas use is reduced by omitting redundant calls and other optimizations. 

min          max            average
60391  ·     131996  ·      123610
54530  ·     125378  ·      117276
(-5,861)     (-6,618)       (-6334)

### CODE REVIEW

**Comment by nxt3d** [comment]
@Arachnid @jefflau I think it might still be possible to get the gas usage even a bit lower. Specifically the _mint function has some redundant calls.

File: contracts/wrapper/NameWrapper.sol [hunk 1]
```diff
@@ -511,27 +511,103 @@ contract NameWrapper is
...
+            // If we pass all the checks, register the subname and wrap it. 
+            bytes memory _name = _saveLabel(parentNode, node, label);
             ens.setSubnodeOwner(parentNode, labelhash, address(this));
-            _wrap(node, name, owner, fuses, expiry);
+
+            // If a owner controlled fuse are being burned, check to make sure 
+            // PARENT_CANNOT_CONTROL and CANNOT_UNWRAP have been burned. 
+            _canFusesBeBurned(node, fuses);
+
+            super._mintWithData(node, owner, fuses, expiry, nodeOwner, nodeFuses, oldExpiry);
```
Comment (nxt3d) [inline contracts/wrapper/NameWrapper.sol:573]: This brings the gas use down to:
min max avarage
60391 · 131996 · 123610
54533 · 124304 · 116263

(-5,858) (-7,692) (-7,347)

**Comment by Arachnid** [comment]
Thank you! Jeff has some changes to setSubnodeOwner to fix a vuln in #159 - can you rebase on top of that?

**Comment by nxt3d** [comment]
> Yes, I reviewed his change. I don't think there are any conflicts.

**Comment by Arachnid** [comment]
Great - can you rebase on top of it anyway?

**Comment by nxt3d** [comment]
> 
I am not sure his change to setSubnodeOwner is necessary. We already check for ens.owner(subnode) in the modifier canCallSetSubnodeOwner. If you look at my version it makes it clear that we check all the conditions, not registered,  registered and not wrapped, and registered and wrapped.

File: contracts/wrapper/NameWrapper.sol [hunk 2]
```diff
@@ -511,27 +517,98 @@ contract NameWrapper is
...
+            // PARENT_CANNOT_CONTROL has not been burned.  
+            if (nodeFuses & PARENT_CANNOT_CONTROL != 0) {
+                revert OperationProhibited(node);
+            }
+
+            //Notice: In order to limit the stack height we are temporaily using label as "name".
+            label = string(_saveLabel(parentNode, node, label));
+
+            // If the name is not wrapped, then register the name and wrap it. 
+            if (nodeOwner == address(0)) {
```
Comment (nxt3d) [inline contracts/wrapper/NameWrapper.sol:594]: ownerOf(uint256(node)) => nodeOwner reduces the gas a lot for the low end: 

My numbers are now: 
min         max          average
60391  ·     131996  ·      123610
53744  ·     124301  ·      116197

(-6,647)      (-7,695)       (-7,413)

File: contracts/wrapper/NameWrapper.sol [hunk 3]
```diff
@@ -511,27 +511,98 @@ contract NameWrapper is
...
+            // If a owner controlled fuse are being burned, check to make sure 
+            // PARENT_CANNOT_CONTROL and CANNOT_UNWRAP have been burned. 
+            _canFusesBeBurned(node, fuses);
+
+            // If we pass all the checks, register the subname and wrap it. 
+            bytes memory _name = _saveLabel(parentNode, node, label);
             ens.setSubnodeOwner(parentNode, labelhash, address(this));
-            _wrap(node, name, owner, fuses, expiry);
+
+            _mintWithDataForUnwrappedNames(node, owner, fuses, expiry, nodeFuses, oldExpiry);
```
Comment (ZhangZhuoSJTU) [inline contracts/wrapper/NameWrapper.sol:573]: Here, assuming the `node` already exists but `baseOwner == address(0)`, i.e., `NameWrapper.ownerOf(node)` is valid. It seems that a malicious parent can re-assign the NameWrapper owner of `node`.

As for how to achieve a crafted state where `baseOwner == address(0)` and `NameWrapper.ownerOf(node)` is valid with fuses burnt, please refer to this [bug report](https://github.com/code-423n4/2022-11-ens-findings/issues/14).

**Comment by ZhangZhuoSJTU** [comment]
Hi, I am sorry that I take it upon myself to jump in the conversation. I am an auditor who participated the last round of ENS C4 auditing and I hope I can provide some help in this case. 

I am not sure whether you can access this [bug report](https://github.com/code-423n4/2022-11-ens-findings/issues/14). The bug is, basically, when we call `setSubnodeOwner` or `setSubnodeRecord` (e.g., `parentNode` and `node` are the parent and child nodes, respectively), it is possible that `ens.owner(node)` is not the NameWrapper contract (or 0), but `NameWrapper.ownerOf(node)` is valid with fuses burnt. Note that _It is a crafted state but not an unwrapped one_. Such vulnerability can induce severe consequence. 

I haven't check the refactoring very carefully (and hence I may say something wrong). But I feel the current refactoring seems to re-introduce the vulnerability mentioned in this [bug report](https://github.com/code-423n4/2022-11-ens-findings/issues/14). I detailed my thinking in the comment and please feel free to let me know there is anything unclear (or I made something wrong).

**Comment by jefflau** [comment]
@nxt3d Have you rebased on top of my fix yet? It would be essential to know if we're introducing any new bugs after patching

**Comment by nxt3d** [comment]
> @nxt3d Have you rebased on top of my fix yet? It would be essential to know if we're introducing any new bugs after patching

I opened another PR with which is the same as this one, except with rebasing in https://github.com/ensdomains/ens-contracts/pull/159

[Refactor of SetSubnodeOwner Including changes made in PR #159](https://github.com/ensdomains/ens-contracts/pull/168)

**Pull request closed without merging**